File: packages/node/src/middleware.ts
import type { RequestHandler } from 'express';
import { errors } from 'jose';
import { verifyLaunchToken } from './verify';
import { consumeSession, readIdentityFields } from './api-client';
import { extractLaunchToken } from './query';
import { InvalidLaunchTokenError, MythosConfigError } from './errors';
import { mythosLog } from './logger';

export function requireLaunchToken(options?: {
  resolveListingIds?: () => Promise<string[]>;
}): RequestHandler {
  return async (req, res, next) => {
    const token = extractLaunchToken(req.query['lt']);
    if (!token) {
      res.status(401).json({ error: 'Missing launch token' });
      return;
    }

    let session;
    try {
      session = await verifyLaunchToken(token, options);
    } catch (err) {
      if (err instanceof MythosConfigError) {
        mythosLog.error('requireLaunchToken: invalid configuration', err);
        res.status(500).json({ error: err.message });
        return;
      }
      if (err instanceof InvalidLaunchTokenError || err instanceof errors.JOSEError) {
        res.status(401).json({ error: 'Invalid launch token' });
        return;
      }
      mythosLog.error('requireLaunchToken: launch token verification failed', err);
      res.status(503).json({ error: 'Could not verify session' });
      return;
    }

    let consumeRes: Response;
    try {
      consumeRes = await consumeSession(session.sessionJti);
    } catch (err) {
      mythosLog.error('requireLaunchToken: consume request failed', err);
      res.status(503).json({ error: 'Could not verify session' });
      return;
    }
    if (consumeRes.status === 409) {
      res.status(401).json({ error: 'Token already consumed' });
      return;
    }
    if (consumeRes.status < 200 || consumeRes.status >= 300) {
      mythosLog.error(`requireLaunchToken: consume returned ${consumeRes.status}`);
      res.status(503).json({ error: 'Could not verify session' });
      return;
    }

    let consumeBody: unknown = null;
    try {
      consumeBody = await consumeRes.json();
    } catch {
      // Identity fields are optional for SDK compatibility; session verification already succeeded.
    }

    req.mythos = { ...session, ...(readIdentityFields(consumeBody) ?? {}) };
    next();
  };
}

File: packages/python/mythos_sdk/middleware.py
from collections.abc import Awaitable, Callable
from dataclasses import replace

from fastapi import HTTPException, Query
import httpx
from jose.exceptions import JOSEError

from .api_client import consume_session, read_identity_fields
from .errors import InvalidLaunchTokenError, MythosConfigError
from .logger import log_error
from .types import MythosSession
from .verify import verify_launch_token


def _attach_identity_token(session: MythosSession, response: httpx.Response) -> MythosSession:
    try:
        identity = read_identity_fields(response.json())
    except (AttributeError, TypeError, ValueError):
        return session
    if identity is None:
        return session
    token, expires_at = identity
    return replace(
        session,
        llmIdentityToken=token,
        llmIdentityExpiresAt=expires_at if isinstance(expires_at, str) else None,
    )


def require_launch_token(
    resolve_listing_ids: Callable[[], Awaitable[list[str]]] | None = None,
) -> Callable[..., Awaitable[MythosSession]]:
    async def dependency(lt: str = Query(..., alias="lt")) -> MythosSession:
        if not lt:
            raise HTTPException(status_code=401, detail="Missing launch token")

        try:
            session = await verify_launch_token(lt, resolve_listing_ids)
        except MythosConfigError as err:
            log_error("require_launch_token: invalid configuration", err)
            raise HTTPException(status_code=500, detail=str(err)) from err
        except (InvalidLaunchTokenError, JOSEError):
            raise HTTPException(status_code=401, detail="Invalid launch token")
        except Exception as err:
            log_error("require_launch_token: launch token verification failed", err)
            raise HTTPException(status_code=503, detail="Could not verify session")

        try:
            resp = await consume_session(session.sessionJti)
        except Exception as err:
            # Network error / unreachable consume endpoint — fail closed, never grant
            # single-use access without a confirmed consume.
            log_error("require_launch_token: consume request failed", err)
            raise HTTPException(status_code=503, detail="Could not verify session")

        if resp.status_code == 409:
            raise HTTPException(status_code=401, detail="Token already consumed")
        if not (200 <= resp.status_code < 300):
            # Any other non-2xx (500, 503, ...) is unconfirmed — fail closed.
            log_error(f"require_launch_token: consume returned {resp.status_code}")
            raise HTTPException(status_code=503, detail="Could not verify session")

        return _attach_identity_token(session, resp)

    return dependency
